tinygrad/codegen/linearizer.py

```diff
@@ -46,6 +46,9 @@ def uop_alu_idx(self, a:UOp, b, ops, ctx:Linearizer, op, dtype=dtypes.int32):
 
   # NOTE: the consts have to be cached for deduping of downstream uops to work
   def const(self, b:Union[int,float], dtype=dtypes.int32, insert_before=None) -> UOp: return self.uop(UOps.CONST, dtype, tuple(), b, insert_before=insert_before)
+  def cast(self, val: UOp, dtype) -> UOp:
+    should_i_cast_you = val.dtype != dtype if self.opts.device != "HIP" else (val.dtype != dtype) or (val.uop == UOps.GEP and val.dtype == dtypes.half and val.vin[0].uop == UOps.LOAD and val.vin[0].dtype == dtypes.half.vec(2) and dtype == dtypes.half)
```

Comment:
Why aren't all calls to cast added to UOps? This line looks sketchy

Reply:
the function is just a utility for cleaner implementations that gets used in the ALU upcast:
https://github.com/tinygrad/tinygrad/pull/2264/files#diff-63d57daaca59e5495bbb108da1c13e782f4a408db8503e015e8094ac07c87ea6R466
This isn't a cast that's explicitly in the AST but we should cast it for compatibility with all backends.

Some backends like Metal don't need all ALU vars to be the same dtype and some (HIP, LLVM) do, so I'm opting to upcast in linearizer.

---
There's another part here, the half2 dtype is defined by HIP's library, but the underlying value of `val0.x` in a half2 isn't half, it's uint16:
https://github.com/argonne-lcf/THAPI/blob/f921880387730e75f04b97a8b07d191a8d28bc28/hip/include/hip/amd_detail/amd_hip_fp16.h#L54-L64

this won't compile:
```
half2 val0 = // ...
half val1 = // ...
val0.x + val1 // error: use of overloaded operator '+' is ambiguous (with operand types 'unsigned short' and 'half' (aka '__half'))
```

I moved this edge-case to linearizer where we do the GEP https://github.com/tinygrad/tinygrad/pull/2264/files#diff-63d57daaca59e5495bbb108da1c13e782f4a408db8503e015e8094ac07c87ea6R118 - so we'll treat the uop as a uint16 just like HIP and leave casting back to half to the other uops if they need to